torchrl/envs/mb_envs/dreamer.py

```diff
@@ -0,0 +1,258 @@
+from math import sqrt
+
+import torch
+import torch.distributions as d
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.distributions.kl import kl_divergence
+
+from torchrl.data import TensorDict
+from torchrl.modules.tensordict_module import TensorDictModule, TensorDictSequence
+from ..model_based import ModelBasedEnv
+
+
+class DreamerEnv(ModelBasedEnv):
+    def __init__(
+        self,
+        obs_depth=32,
+        rssm_hidden=200,
+        rnn_hidden_dim=200,
+        state_dim=20,
+        device="cpu",
+        dtype=None,
+        batch_size=None,
+    ):
+        super().__init__(
+            world_model=DreamerModel(
+                obs_depth=obs_depth,
+                rssm_hidden=rssm_hidden,
+                rnn_hidden_dim=rnn_hidden_dim,
+                state_dim=state_dim,
+            ),
+            reward_model=TensorDictModule(
+                RewardModel(),
+                in_keys=[
+                    "posterior_states",
+                    "beliefs",
+                ],
+                out_keys=["predicted_reward"],
+            ),
+            device=device,
+            dtype=dtype,
+            batch_size=batch_size,
+        )
+
+        self.inference_world_model = self.word_model.select_subsequence(
+            in_keys=["initial_state", "initial_rnn_hidden", "action"],
+            out_keys=["prior_states", "prior_rnn_hiddens"],
+        )
+        self.inference_reward_model.in_keys = [
+            "prior_states",
+            "beliefs",
+        ]
+
+    def _set_optimizer(self) -> torch.optim.Optimizer:
+        return torch.optim.Adam(
+            list(self.word_model.parameters()) + list(self.reward_model.parameters()),
+            lr=1e-3,
+            eps=1e-4,
+        )
+
+    def loss(self, tensordict: TensorDict) -> torch.Tensor:
+        B, T, D = tensordict["prior_means"].shape
+        flat_prior_means = tensordict["prior_means"].reshape(B * T, D)
```

Comment:
It's the typical case where I would use tensordict.view(T*B, -1)
And then get("key1") and "key2"